notebooks/byod.py

```diff
@@ -98,27 +98,33 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, _ = filename.split(".", 1)                                                                    
+        if filename.endswith(".cram"): 
+            ext = "cram" 
+        elif filename.endswith(".crai") or filename.endswith(".cram.crai"):
+            ext = "crai" 
+        else:
+            # Metadata etc often takes the form of *.txt files so we'll ignore
+            # it rather than raising an exception
+            print(f"Unable to parse {filename} into sample and extension")
+        return sample, ext
+
     def create_cram_crai_table(table: str, listing: Iterable[str]):
         crams = dict()
         crais = dict()
         for key in listing:
             _, filename = key.rsplit("/", 1)
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
+            ext = parse_cram_crai(filename)
             if "cram" == ext:
                 crams[sample] = key
             elif "crai" == ext:
                 crais[sample] = key
             else:
                 continue
         samples = sorted(crams.keys())
+        if len(crams) != len(crais):
+            raise KeyError("Mismatch in number of cram and crai files.")
```

Comment:
This isn't really a key error. Instead, consider:
```
assert len(crams) == len(crais), "Mismatch in number of cram and crai files."
```

Alternatively, we could find matched and unmatched crams and crais:
```
common_samples = set(crams.keys()).intersection(set(crais.keys())
unmatched_crams = set(crams.keys()).difference(set(crais.keys()))
unmatched_crais = set(crais.keys()).difference(set(crams.keys()))
if unmatched_crams:
    warnings.warn("Crams missing crais for samples {unmatched_crams}")
if unmatched_crais:
    warnings.warn("Crais missing crams for samples {unmatched_crais}")
```
and construct the table from `common_samples`.